src/components/pdf-export/pdf-export.service.ts

```diff
@@ -0,0 +1,150 @@
+import { Injectable, Renderer2 } from '@angular/core';
+import { PdfExportItemDirective } from './pdf-export-item.component';
+
+@Injectable()
+export class PdfExportService {
+
+    private _items: PdfExportItemDirective[] = [];
+
+    registerItem(item: PdfExportItemDirective): void {
+        this._items.push(item);
+    }
+
+    unregisterItem(item: PdfExportItemDirective): void {
+        let idx = this._items.indexOf(item);
+
+        if (idx !== -1) {
+            this._items.splice(idx, 1);
+        }
+    }
+
+    getDocument(): HTMLHtmlElement {
+
+        // group items by row
+        let groupings: PDFExportGrouping[] = [];
+
+        this._items.forEach(item => {
+
+            let row = groupings.find(group => group.row === item.row);
+
+            if (!row) {
+                groupings.push({ row: item.row, items: [{ columns: item.columns, element: this.clone(item.getElement()) }] });
+            } else {
+                row.items.push({ columns: item.columns, element: this.clone(item.getElement()) });
+            }
+
+        });
+
+        // map to an html structure
+        let rows = groupings.map(rowData => {
+            let row = document.createElement('div');
+            row.classList.add('row');
+
+            rowData.items.forEach(item => {
+                let column = document.createElement('div');
+
+                if (item.columns) {
+                    column.classList.add(`col-md-${item.columns}`);
+                }
+
+                column.appendChild(item.element);
+                row.appendChild(column);
+            });
+
+            return row;
+        });
+
+ 
+        let head = this.cloneHead();
+        let body = document.createElement('body');
+        let htmlElement = document.createElement('html');
+
+        rows.forEach(row => {
+            body.appendChild(row);
+        });
+
+        htmlElement.appendChild(head);
+        htmlElement.appendChild(body);
+
+        return htmlElement;
+    }
+
+    private cloneHead(): HTMLHeadElement {
+        let head = document.head.cloneNode(true) as HTMLHeadElement;
+        return head;
+    }
+
+    private clone(element: HTMLElement): HTMLElement {
+        return element instanceof HTMLCanvasElement ? this.cloneCanvas(element) 
```

Comment:
Change this to a if statement. Try to avoid nested ternary if statements as they get hard to follow very quickly.